File: numbers-to-words/python/src/numbers_to_words/numbers_to_words.py
_ONES = (
    "zero", "one", "two", "three", "four",
    "five", "six", "seven", "eight", "nine",
    "ten", "eleven", "twelve", "thirteen", "fourteen",
    "fifteen", "sixteen", "seventeen", "eighteen", "nineteen",
)

_TENS = (
    "", "", "twenty", "thirty", "forty",
    "fifty", "sixty", "seventy", "eighty", "ninety",
)


def to_words(n: int) -> str:
    if n == 0:
        return "zero"

    parts: list[str] = []

    thousands, n = divmod(n, 1000)
    if thousands:
        parts.append(f"{_ONES[thousands]} thousand")

    hundreds, n = divmod(n, 100)
    if hundreds:
        parts.append(f"{_ONES[hundreds]} hundred")

    if n:
        parts.append(_below_hundred(n))

    return " ".join(parts)


def _below_hundred(n: int) -> str:
    if n < 20:
        return _ONES[n]
    tens, ones = divmod(n, 10)
    return _TENS[tens] if ones == 0 else f"{_TENS[tens]}-{_ONES[ones]}"

File: numbers-to-words/typescript/src/numbersToWords.ts
const ONES = [
  'zero', 'one', 'two', 'three', 'four',
  'five', 'six', 'seven', 'eight', 'nine',
  'ten', 'eleven', 'twelve', 'thirteen', 'fourteen',
  'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen',
] as const;

const TENS = [
  '', '', 'twenty', 'thirty', 'forty',
  'fifty', 'sixty', 'seventy', 'eighty', 'ninety',
] as const;

export function toWords(n: number): string {
  if (n === 0) return 'zero';

  const parts: string[] = [];

  const thousands = Math.floor(n / 1000);
  if (thousands > 0) {
    parts.push(`${ONES[thousands]} thousand`);
    n %= 1000;
  }

  const hundreds = Math.floor(n / 100);
  if (hundreds > 0) {
    parts.push(`${ONES[hundreds]} hundred`);
    n %= 100;
  }

  if (n > 0) {
    parts.push(belowHundred(n));
  }

  return parts.join(' ');
}

function belowHundred(n: number): string {
  if (n < 20) return ONES[n]!;
  const tens = Math.floor(n / 10);
  const ones = n % 10;
  return ones === 0 ? TENS[tens]! : `${TENS[tens]}-${ONES[ones]}`;
}
